packages/manager/src/layouts/OAuth.tsx

```diff
@@ -1,149 +1,207 @@
+    this.exchangeAuthorizationCodeForToken(code, returnTo, nonce);
+  }
 
-    /**
-     * We have all the information we need and can persist it to localStorage and Redux.
-     */
-    this.props.dispatchStartSession(
-      accessToken,
-      tokenType,
-      scopes,
-      expireDate.toString()
-    );
+  createFormData(
+    clientID: string,
+    code: string,
+    nonce: string,
+    codeVerifier: string
+  ): FormData {
+    const formData = new FormData();
+    formData.append('grant_type', 'authorization_code');
+    formData.append('client_id', clientID);
+    formData.append('code', code);
+    formData.append('state', nonce);
+    formData.append('code_verifier', codeVerifier);
+    return formData;
+  }
 
-    /**
-     * All done, redirect this bad-boy to the returnTo URL we generated earlier.
-     */
-    history.push(returnTo);
+  async exchangeAuthorizationCodeForToken(
+    code: string,
+    returnTo: string,
+    nonce: string
+  ) {
+    try {
+      const expireDate = new Date();
+      const codeVerifier = authentication.codeVerifier.get();
+
+      if (codeVerifier) {
+        authentication.codeVerifier.set('');
+
+        /**
+         * We need to validate that the nonce returned (comes from the location query param as the state param)
+         * matches the one we stored when authentication was started. This confirms the initiator
+         * and receiver are the same.
+         */
+
+        this.checkNonce(nonce);
+
+        const formData = this.createFormData(
+          `${clientID}`,
+          code,
+          nonce,
+          codeVerifier
+        );
+
+        this.setState({ isLoading: true });
+
+        const response = await fetch(`${loginURL}/oauth/token`, {
+          body: formData,
+          method: 'POST',
+        });
+
+        this.setState({ isLoading: false });
+
+        if (response.ok) {
+          const tokenParams = await response.json();
+
+          /**
+           * We multiply the expiration time by 1000 ms because JavaSript returns time in ms, while
+           * the API returns the expiry time in seconds
+           */
+
+          expireDate.setTime(
+            expireDate.getTime() + +tokenParams.expires_in * 1000
+          );
+
+          this.props.dispatchStartSession(
+            tokenParams.access_token,
+            tokenParams.token_type,
+            tokenParams.scopes,
+            expireDate.toString()
+          );
+
+          /**
+           * All done, redirect this bad-boy to the returnTo URL we generated earlier.
+           */
+          this.props.history.push(returnTo);
+        } else {
+          clearStorageAndRedirectToLogout();
+        }
+      } else {
+        clearStorageAndRedirectToLogout();
+      }
+    } catch (error) {
+      clearStorageAndRedirectToLogout();
+    }
   }
 
   render() {
+    const { isLoading } = this.state;
+
+    if (isLoading) {
+      return <CircleProgress />;
```

Comment:
Can you use `SplashScreen ` here instead of `CircleProgress`. Using `CircleProgress ` is causing the layout to shift because it doesn't use the entire height of the page

```suggestion
      return <SplashScreen />;
```

Reply:
Thanks. Updated.